scaling-puppet-enterprise.md

```diff
@@ -0,0 +1,55 @@
+# Scaling Puppet Enterprise
+
+## Summary
+
+This standard defines how to determine bottlenecks in PE and think about adding compile masters or adding resources to an existing PE installation.
+
+## Expectations
+
+This standard relies on the [Monitoring PE standard]() and assumes the user will have metrics that can be trended over time.  
+
+There is also an expectation that the user will have already performed tuning to have the correct amount of heap, number of JRubies, command processing threads, etc… Tuning should be completed before scaling with additional hardware.  For simplicity tuning and scaling are treated as separate topics, we assume you already tuned appropriately then metrics can indicate the need for more hardware resources.
+
+## Standard Details
+
+### Preferred Option
+
+There are three main bottlenecks in PE:
+
+Catalog compilation
+PostgreSQL ( the backend database to PuppetDB )
+PuppetDB service
+
+You should assume that you will run into a bottleneck in catalog compilation before you will run into the need to scale PostgreSQL.  After you have added compile masters to address the need for more catalog compilation then you may need to add resources to the node serving as your PostgreSQL instance.  Finally, after a sufficient amount of load has been reached you may reach a bottleneck in the PuppetDB service’s ability to process commands and place them in PostgreSQL.
+
+#### Determining the need to add compile masters
+
+Avg free JRubies is the primary metric to use when determining when you need to add compile masters.  When you deploy new nodes, the initial pluginsync they need to perform will cause an increased need for JRubies that should then reduce after they have received the initial pluginsync.  Load avg and avg free JRubies should be tied at the hip except for when you are doing that initial pluginsync in which case avg free JRubies will be near 0 and you will not be maxing out the load. After steady state is achieved and most JRubies are compiling catalogs then avg free JRubies and load avg will be tightly correlated again.  Based on the number of compile masters you have you can reasonably estimate the number of nodes a compile master must accommodate.  Assume you have 10K nodes and 3 compile masters with 1 avg free JRuby on each of the compile masters.  
+
+10K / 3 = 3,333 nodes per compile master 
+
+The magic number for how many nodes a compile master can support changes over time as puppet code is added, catalogs change, and/or facts are added.  When you are running compile masters with more than 1 avg free JRuby then you have capacity to add more nodes to your PE installation.  Another useful calculation is nodes per JRuby.
```

Comment:
@npwalker The second sentence sounds a little weird. Can you put it another way as I am a little confused.

Reply:
@npwalker can we add something like 'Puppet recommends an average of one free jruby per compile masters. If adding new nodes you should be concerned if this average is less than one'?